rs/execution_environment/src/execution_environment.rs

```diff
@@ -2111,6 +2111,7 @@ impl ExecutionEnvironment {
 
         let result = match result {
             Ok(new_canister) => {
+                state.metadata.heap_delta_estimate += NumBytes::from(new_canister.memory_usage());
```

Comment:
I think this is the wrong number to add to heap delta. This way restoring a canister with 140GB+ stable memory will stall the subnet until the next checkpoint. Which is completely different from what the heap delta usually measures, namely new changes since the last checkpoint.

What we actually should limit is the amount of data that needs to be persisted. I suggest to use [PageMap::page_delta](https://github.com/dfinity/ic/blob/f116e57138d823d5a3a4302810b4aa6ec7fb32f8/rs/replicated_state/src/page_map.rs#L442) instead, particularly `page_delta.len() * PAGE_SIZE`. 

The same delta should also be applied when creating a new snapshot.

Reply:
What I wasn't sure when loading a snapshot is whether we're clever enough to know what's the diff between current stable memory and the one we're restoring to. It seems that you're saying the `PageMap` will know this. I can update the number for sure.

> The same delta should also be applied when creating a new snapshot.

We do apply the `new_snapshot_size`, [reference](https://sourcegraph.com/github.com/dfinity/ic@f116e57138d823d5a3a4302810b4aa6ec7fb32f8/-/blob/rs/execution_environment/src/canister_manager.rs?L1923). I guess that's incorrect too? If yes, then I'll repurpose the PR to fix both of these.

Reply:
> We do apply the new_snapshot_size, [reference](https://sourcegraph.com/github.com/dfinity/ic@f116e57138d823d5a3a4302810b4aa6ec7fb32f8/-/blob/rs/execution_environment/src/canister_manager.rs?L1923). I guess that's incorrect too?

Yes the same point applies there. In fact, that direction is the important one, because on restore we usually don't have to do any persisting, but just need to hardlink some files around. This is because if there was a checkpoint between backup and restore, we would have done the persisting then.

>What I wasn't sure when loading a snapshot is whether we're clever enough to know what's the diff between current stable memory and the one we're restoring to.

Not quite. We are however clever enough to know the difference between the files we have in the snapshot folder and the actual state of the PageMap, and it's exactly that `page_delta`. (Actually it's `unflushed_page_delta`, but I would consider that an implementation detail and rather overcount a bit).

Reply:
Yes, but also don't forget the wasm chunk store. I am not quite sure if other non-PageMap things also go into the heap delta usually, but if they do then you might want to add them here too.

Reply:
I don't think other non-PageMap things matter because as you said we're interested in the ones that result in something that will need to be persisted to disk. Also, the heap_delta reported from normal message executions covers only wasm and stable memory iirc (wasm chunk store APIs update heap delta directly since it's in the management canister as well).

If I move this function to `CanisterState` I can capture everything (wasm, stable and chunk store).

Reply:
Hmmm, thinking a bit more about it, I'm not sure if using the amounts from the canister state would be correct.

Imagine the following scenario:
1. Checkpoint happened.
2. Take a snapshot - creates some `heap_delta1`.
2. Perform some changes to the canister - creates some `heap_delta2`.
3. Take another snapshot (given our current policy the previous one is deleted) - creates some `heap_delta3`.
4. Another checkpoint happens.

With the current approach we would be summing up all heap delta. Is this really the right thing to do? Isn't `heap_delta3` a superset of `heap_delta1`? And `heap_delta1` won't actually get persisted, right?

Reply:
You are right we are overcounting in this case, but very much in line of how much we overcount heap deltas generally. For example this scenario:

Message 1: canister writes to page 0 of its heap
Message 2: canister overwrites page 0 of its heap
....
Message n: canister overwrites page 0 of its heap
Checkpoint happens

In this case, we in fact accumulated `n` pages worth of heap delta, but end up persisting only one.

Also in your example, whether the first snapshot got persisted then deleted or just never persisted is an implementation detail of the state manager and in fact we might do either, depending on what round this happens at exactly.

Reply:
Pushed the changes, ptal.